### PR TITLE
Fix dictionary key sorting for JAX PyTrees

### PR DESCRIPTION
**Pull request by XuehaiPan**
Part of google/jax#11871

Function `list.sort` does not support sorting with different types.

```python
>>> sorted({1: 0, 'a': 1})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'str' and 'int'
```

This will cause errors when manipulating dictionaries that have different key types.

```python
>>> import jax.tree_util as pytree
>>> pytree.tree_flatten({1: 1, 2: 2})
([1, 2], PyTreeDef({1: *, 2: *}))
>>> pytree.tree_flatten({'a': 1, 'b': 2})
([1, 2], PyTreeDef({'a': *, 'b': *}))
>>> pytree.tree_flatten({1: 1, 'a': 2})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../venv/lib/python3.9/site-packages/jax/_src/tree_util.py", line 60, in tree_flatten
    return pytree.flatten(tree, is_leaf)
TypeError: '<' not supported between instances of 'str' and 'int'
```

Since Python 3.6, the keys are insertion ordered.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @XuehaiPan Can you please resolve conflicts? Thank you!

File: tensorflow/compiler/xla/python/pytree.cc [hunk 1]
```diff
@@ -90,27 +90,50 @@ namespace py = pybind11;
...
+          // The keys remain in the insertion order.
+          PyErr_Clear();
+        } else [[unlikely]] {
+          throw;
+        }
+      }
+    } else [[unlikely]] {
+      throw;
+    }
+  }
```
Comment (XuehaiPan) [inline tensorflow/compiler/xla/python/pytree.cc:129]: Python equivalent code:

```python
def get_sorted_dict_keys(dct: dict[KT, VT]) -> list[KT]:
    keys = list(dct)
    try:
        # Sort directly if possible (do not use `key` for performance reasons)
        return sorted(keys)
    except TypeError:  # the keys are not comparable
        try:
            # Add `{obj.__class__.__module__}.{obj.__class__.__qualname__}` to the key order to make
            # it sortable between different types (e.g. `int` vs. `str`)
            return sorted(
                keys,
                # pylint: disable-next=consider-using-f-string
                key=lambda o: ('{0.__module__}.{0.__qualname__}'.format(o.__class__), o),
            )
        except TypeError:  # cannot sort the keys (e.g. user-defined types)
            return keys  # fallback to insertion order
```

**Comment by XuehaiPan** [comment]
> Hi @XuehaiPan Can you please resolve conflicts? Thank you!

@gbaned Done.

File: tensorflow/compiler/xla/python/pytree.cc [hunk 2]
```diff
@@ -90,27 +90,50 @@ namespace py = pybind11;
...
-  std::stable_sort(keys.begin(), keys.end(),
-                   [&ret](const py::object& a, const py::object& b) {
-                     int cmp =
-                         PyObject_RichCompareBool(a.ptr(), b.ptr(), Py_LT);
-                     if (cmp == -1) ret = -1;
-                     return cmp;
-                   });
-  if (ret == -1) {
-    throw py::error_already_set();
-  }
```
Comment (XuehaiPan) [inline tensorflow/compiler/xla/python/pytree.cc:112]: The old implementation is `py::list -> std::vector -> std::sort`. Changed to `py::list -> py::sort -> std::vector` for better Python exception handling.

File: tensorflow/compiler/xla/python/pytree.cc [hunk 3]
```diff
@@ -90,27 +90,50 @@ namespace py = pybind11;
...
+          // The keys remain in the insertion order.
+          PyErr_Clear();
+        } else {
+          throw;
+        }
+      }
+    } else {
+      throw;
+    }
+  }
```
Comment (XuehaiPan) [inline tensorflow/compiler/xla/python/pytree.cc:129]: Python equivalent code:

```python
def get_sorted_dict_keys(dct: dict[KT, VT]) -> list[KT]:
    keys = list(dct)
    try:
        # Sort directly if possible (do not use `key` for performance reasons)
        return sorted(keys)
    except TypeError:  # the keys are not comparable
        try:
            # Add `{obj.__class__.__module__}.{obj.__class__.__qualname__}` to the key order to make
            # it sortable between different types (e.g. `int` vs. `str`)
            return sorted(
                keys,
                # pylint: disable-next=consider-using-f-string
                key=lambda o: ('{0.__module__}.{0.__qualname__}'.format(o.__class__), o),
            )
        except TypeError:  # cannot sort the keys (e.g. user-defined types)
            return keys  # fallback to insertion order
```

**Comment by sherhut** [comment]
We can only land this if the dependent change in JAX also lands if I am not mistaken. Can we keep this a draft until the other change is accepted @XuehaiPan ?

**Comment by XuehaiPan** [comment]
Convert to draft PR and would wait for feedback in the downstream JAX repo.

**Comment by gbaned** [comment]
Hi @XuehaiPan Any update on this PR? Please. Thank you!

**Comment by XuehaiPan** [comment]
> Hi @XuehaiPan Any update on this PR? Please. Thank you!

Hi @gbaned, we would wait until the proposed change is also accepted by the JAX downstream.

**Comment by gbaned** [comment]
Hi @XuehaiPan Any update on this CL? Please. Thank you!

**Comment by XuehaiPan** [comment]
Still, no feedback has been received from the JAX downstream yet. It appears that this change might not be under consideration, as the JAX community might prefer raising errors for unsortable keys. I propose closing this PR for now and continuing the discussion within the JAX repo.

**Pull request closed without merging**